source/slang/slang-check-expr.cpp

```diff
@@ -2185,6 +2185,76 @@ namespace Slang
         return _canLValueCoerceScalarType(a, b);
     }
 
+
+    void SemanticsVisitor::compareMemoryQualifierOfParamToArgument(
+        ParamDecl* paramIn,
+        Expr* argIn)
+    {
+        auto arg = as<VarExpr>(argIn);
+        if (!paramIn || !arg)
+            return;
+
+        auto argDeclRef = arg->declRef;
+        if (!argDeclRef)
+            return;
+
+        ASTNodeType opThis;
+        ASTNodeType foundOp;
+        for (auto modThis : argDeclRef.getDecl()->modifiers)
+        {
+            UnownedStringSlice target;
+            opThis = modThis->astNodeType;
+            foundOp = ASTNodeType::NodeBase;
+            if (opThis == ASTNodeType::GloballyCoherentModifier)
+            {
+                foundOp = opThis;
+                target = UnownedStringSlice("coherent");
+            }
+            else if (opThis == ASTNodeType::GLSLVolatileModifier)
+            {
+                foundOp = opThis;
+                target = UnownedStringSlice("volatile");
+            }
+            else if (opThis == ASTNodeType::GLSLRestrictModifier)
+            {
+                foundOp = opThis;
+                target = UnownedStringSlice("restrict");
+            }
+            if (opThis == ASTNodeType::GLSLReadOnlyModifier)
+            {
+                foundOp = opThis;
+                target = UnownedStringSlice("readonly");
+            }
+            else if (opThis == ASTNodeType::GLSLWriteOnlyModifier)
+            {
+                foundOp = opThis;
+                target = UnownedStringSlice("writeonly");
+            }
+
+            if (foundOp == ASTNodeType::NodeBase)
+                continue;
+
+            // Only texture types are allowed to have memory qualifiers on parameters
+            if(paramIn->getType()->astNodeType != ASTNodeType::TextureType)
```

Comment:
This can be moved to `SemanticsVisitor::checkModifiers`, otherwise we are repeating this check on every call.

As part of that checking, you can fold all image modifiers into a combined flag in the form of a `CheckedImageModifiers` Modifier, whose content is a combined uint bits:

```
class CheckedGLSLImageMemoryModifiers: Modifier
{
     SLANG_AST_CLASS(CheckedGLSLImageModifiers)
     GLSLImageMemoryModifierBit combinedBits;
}
```

So if we check `coherent writeonly`, we will fold those two modifiers into one `CheckedGLSLImageMemoryModifiers` that is `GLSLImageMemoryModifierBit::Coherent | GLSLImageMemoryModifierBit::WriteOnly`

Reply:
You can modifier `checkModifier` to return a `CheckedGLSLImageMemoryModifiers` on a `GLSL**MemoryModifier`,
and combine them in `checkModifier` into a single `CheckedGLSLImageMemoryModifiers`.

Reply:
After checking the modifiers and folding them  into `CheckedGLSLImageMemoryModifiers`, you can add checking logic in `SemanticsDeclBodyVisitor::visitParamDecl` to issue the `image memory modifier on non-iimage type` diagnostic.